tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/Emitter.scala

```diff
@@ -66,32 +79,72 @@ final class Emitter private (semantics: Semantics, outputMode: OutputMode,
     emitLines(customHeader, builder)
 
   def emitPrelude(builder: JSFileBuilder, logger: Logger): Unit = {
-    outputMode match {
-      case OutputMode.ECMAScript51Global =>
-      case OutputMode.ECMAScript51Isolated | OutputMode.ECMAScript6 =>
-        builder.addLine("(function(){")
+    moduleKind match {
+      case ModuleKind.NoModule =>
+        outputMode match {
+          case OutputMode.ECMAScript51Global =>
+          case OutputMode.ECMAScript51Isolated | OutputMode.ECMAScript6 =>
+            builder.addLine("(function(){")
+        }
+
+      case ModuleKind.CommonJSModule =>
     }
 
     builder.addLine("'use strict';")
-    builder.addFile(CoreJSLibs.lib(semantics, outputMode))
+    builder.addFile(CoreJSLibs.lib(semantics, outputMode, moduleKind))
   }
 
   def emit(unit: LinkingUnit, builder: JSTreeBuilder, logger: Logger): Unit = {
     startRun(unit)
     try {
       val orderedClasses = unit.classDefs.sortWith(compareClasses)
+
+      emitModuleImports(orderedClasses, builder)
+
       for (classInfo <- orderedClasses)
         emitLinkedClass(classInfo, builder)
     } finally {
       endRun(logger)
     }
   }
 
+  private def emitModuleImports(orderedClasses: List[LinkedClass],
+      builder: JSTreeBuilder): Unit = {
+    moduleKind match {
+      case ModuleKind.NoModule =>
+
+      case ModuleKind.CommonJSModule =>
+        val jsDesugaring = new JSDesugaring(internalOptions)
+        val encounteredModuleNames = mutable.Set.empty[String]
+        for (classDef <- orderedClasses) {
+          classDef.jsNativeLoadSpec match {
+            case None =>
+            case Some(JSNativeLoadSpec.Global(_)) =>
+
+            case Some(JSNativeLoadSpec.Import(module, _)) =>
+              if (encounteredModuleNames.add(module)) {
+                implicit val pos = classDef.pos
+                val rhs = js.Apply(js.VarRef(js.Ident("require")),
+                    List(js.StringLiteral(module)))
+                val lhs = jsDesugaring.envModuleField(module)
+                val decl = jsDesugaring.genLet(lhs.ident, mutable = false, rhs)
+                builder.addJSTree(decl)
+              }
+          }
+        }
+    }
+  }
+
   def emitPostlude(builder: JSFileBuilder, logger: Logger): Unit = {
-    outputMode match {
-      case OutputMode.ECMAScript51Global =>
-      case OutputMode.ECMAScript51Isolated | OutputMode.ECMAScript6 =>
-        builder.addLine("}).call(this);")
+    moduleKind match {
+      case ModuleKind.NoModule =>
+        outputMode match {
+          case OutputMode.ECMAScript51Global =>
+          case OutputMode.ECMAScript51Isolated | OutputMode.ECMAScript6 =>
+            builder.addLine("}).call(this);")
+        }
+
+      case ModuleKind.CommonJSModule =>
```

Comment:
Consider moving the logic of whether you need an isolation guard to a separate method.